2795 s1 duplicates

elipe17 opened the pull request:
## Summary of Changes
- Added new class that manages duplicate detection while case cat4 is occurring
- Updated cat4 validation to leverage duplicate detector
- Updated parse logic to not create records in a case if it has any cat4 errors
Pull request closes #2795 

## CPU and Memory Analysis
- Used Docker to track the `web` container stats during parsing of the largest file we have in the repo (~50MB) on the `develop` branch and this branch. 
- The overall results can be found in `super_big_file_results.txt`. The data listed in that text file can be found in the accompanying files: `develop_super_big_file.txt` and `dup_super_big_file.txt`.
- The data was generated with: `track_docker_stats.sh.txt` (GitHub won't let you upload a pure shell file).
[super_big_file_results.txt](https://github.com/user-attachments/files/15534179/super_big_file_results.txt)
[dup_super_big_file.txt](https://github.com/user-attachments/files/15534181/dup_super_big_file.txt)
[develop_super_big_file.txt](https://github.com/user-attachments/files/15534183/develop_super_big_file.txt)
[track_docker_stats.sh.txt](https://github.com/user-attachments/files/15534202/track_docker_stats.sh.txt)


## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
2. Create a file for TANF section 1 - 4 that has one or more duplicate or partial duplicate records.
3. Submit file
4. Verify no records tied to that case are created and the error report reflects cat4 duplicate errors

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] In section 1 files, exact and partial duplicate records are detected, not stored in db, and error message included in the feedback report.
- [x] In section 2 files, exact and partial duplicate records are detected, not stored in db, and error message included in the feedback report.
- [x] section 3 files, partial duplicate T6 records are detected, not stored in db, and error message included in the feedback report.
- [x] section 4 files, partial duplicate T7 records are detected, not stored in db, and error message included in the feedback report.
- [x] duplicate detection logic is applied consistently to TANF, SSP, and Tribal TANF files. 
- [x] cat 4 records not stored in db
- [ ] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented?

elipe17 committed:
- Initial commit for duplicate record manager for TANF section 1 records

elipe17 committed:
- more efficient error generation

elipe17 committed:
- Added a way to track precedence in the error messages to avoid generating less meaningful messages
- Updated error generation

elipe17 committed:
-  Fix lint
- Add docstrings

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
- Added support to delete deplicate records

elipe17 committed:
- add support for bulk/raw delete

elipe17 committed:
- Update deletes on rollbacks to raw delete since it wont hurt anything

elipe17 committed:
- Updated elastic and django to support raw/bulk deleting

elipe17 committed:
- Updating rollback test since it is very relevant now that we are doing raw deletes

elipe17 committed:
- raise log level

elipe17 committed:
- Update test

elipe17 committed:
- remove import

elipe17 committed:
- Update error generation to use record and schema
- rename var

elipe17 committed:
- Update error precedence logic

elipe17 committed:
- fix lint

elipe17 committed:
Merge branch '2842-cat-4-remaining-s2-validators' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
- use correct method

elipe17 committed:
- remove extra call to bulk create errors

elipe17 committed:
- move duplicate manager into case consistency validator
- Starting to remove SortedRecordSchemaPairs from case consistency validator
- updating to support in memory record removal if they havent been serialized

elipe17 committed:
- update add_record logic to include rpt_month_year

elipe17 committed:
- Move SortedRecordSchema pairs to util
- Update case consistency validator to not use OG SortedRecordSchemaPairs
- Update dup logic to not consider records on the same line

elipe17 committed:
- remove unused import

elipe17 committed:
- Update class to store records based on hash pending the section type

elipe17 committed:
- Update parse to leverage new data structure

elipe17 committed:
- clearing after bulk create

elipe17 committed:
- Rename to be more accurate
- Add method to generate bulk create dictionary
- Stub function for removing cases

elipe17 committed:
- Updated class name
- basing off of doc instead of schema

elipe17 committed:
- Adding infrastructure to remove records from memory or the DB pending their state

elipe17 committed:
- fix logic issue

elipe17 committed:
- fix logic error

elipe17 committed:
- added back list of cases

elipe17 committed:
- fixed test_add_record

elipe17 committed:
- Fix all case consistency tests

elipe17 committed:
- Fix most lint errors

elipe17 committed:
- remove print

elipe17 committed:
- update rollback logic
- fix most parse tests

elipe17 committed:
- Update duplicate error messages
- Added logging
- Updated tests

elipe17 committed:
- functionized dup manager a bit
- updated validator to query errors from dup manager
- added test for s1 duplicate detection

elipe17 committed:
- Add test for S1 partial duplicate detection

elipe17 committed:
- fix lint

elipe17 committed:
- Add section 2 tests

elipe17 committed:
- Fix reference error

elipe17 committed:
- order by id

elipe17 committed:
- parametrizing batch size
- initial integration tests

elipe17 committed:
- move creation logic in parse.py
- add container to monitor what cases have serialized records
- add test for in memory and DB based record removal with dupes detected

elipe17 committed:
- added section 2 duplicate tests

elipe17 committed:
- Test for partial duplicates
- test doesnt behave as I would expect

elipe17 committed:
- fix lint

elipe17 committed:
- Fixed logic error where last case in file wouldnt be removed if it has cat4 errors

elipe17 committed:
- Update SortedRecords to leverage an unsorted container to make bulk  createing faster. Dont have to iterate/generate the required structure

elipe17 committed:
- naming and doc string updates

elipe17 committed:
- renaming functions
- removing unnecessary members

elipe17 committed:
- update doc strings

elipe17 committed:
- Add test for family affiliation negating partial duplicity

elipe17 committed:
- Updated to support duplicate checking on section 3/4

elipe17 committed:
- fix lint

elipe17 committed:
- Update docstring

jtimpe commented on tdrs-backend/tdpservice/parsers/case_consistency_validator.py:
```diff
@@ -187,77 +208,74 @@ def __validate_s1_records_are_related(self):
                         msg=(
-                            f'Every {t2_model_name} record should have at least one corresponding '
-                            f'{t1_model_name} record with the same RPT_MONTH_YEAR and CASE_NUMBER.'
+                            f'There should only be one {t1_model_name} record '
```
~~i don't believe this error gets triggered anymore~~ - the error gets generated but probably shouldn't. could you remove the "at least one" t1/t4 errors? then as discussed in OH, potentially don't cat4 validate when a record is a duplicate

elipe17 committed:
Merge branch '2842-cat-4-remaining-s2-validators' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
- Update to allow all record types to have duplicate detection

elipe17 committed:
- add duplicate detection unit tests for all program types and record types

elipe17 committed:
- fix lint

elipe17 committed:
- Add cases for ssp

elipe17 committed:
- Move parser fixtures to their own conftest.py

elipe17 committed:
- Move conftest.py to the test folder

elipe17 committed:
- removing whitespace

jtimpe commented on tdrs-backend/tdpservice/parsers/case_consistency_validator.py:
```diff
@@ -77,31 +49,80 @@ def __generate_and_add_error(self, schema, record, field, msg):
+
+        self.current_hash = hash_val
+
+        # TODO: Duplicate detection applies to all sections, however we might need to implement a factory of some sort
```
still relevant?

elipe17 replied:
No sir!

jtimpe commented on tdrs-backend/tdpservice/parsers/duplicate_manager.py:
```diff
@@ -0,0 +1,203 @@
+    def __skip_partial(self, record):
+        skip_partial = False
+        if record.RecordType == "T2":
+            skip_partial = record.FAMILY_AFFILIATION in {3, 5}
```
i'm a little concerned that this logic is buried pretty deep within parsing/validation (both in terms of runtime logic and code/file structure) but makes direct reference to possible values for a specific field in specific schemas (some of which are sort of not captured here, either, becuase of how `RecordType` is used). I wonder if there's a better place to put this, perhaps on the schema itself? this may be better left as a future note as i could see wanting to explore some different options.

elipe17 replied:
- Because duplicate detection is program type agnostic we should be able to move this check up into the schema itself to alleviate the testing required to hit all these branches.

jtimpe reviewed (changes requested):
only requested change is the removal of the old duplicate validator. otherwise, a couple questions. everything else is looking really good!

elipe17 commented on tdrs-backend/docker-compose.yml:
```diff
@@ -102,6 +102,8 @@ services:
       - GENERATE_TRAILER_ERRORS=True
       - BYPASS_KIBANA_AUTH
+      - IGNORE_DUPLICATE_ERROR_PRECEDENCE
+      - BULK_CREATE_BATCH_SIZE
```
Move these into the test files themselves

raftmsohani commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -174,3 +174,67 @@ def get_years_apart(rpt_month_year_date, date):
+
+            hashed_case[document] = records
+            self.hash_sorted_cases[hash_val] = hashed_case
+            # We treat the nested dictionary here as a set because dictionaries are sorted while sets aren't. If we
```
If you mean dict automatically sorts on keys, I don't think that is true

elipe17 replied:
As of Python 3.7 dictionaries are [ordered](https://docs.python.org/3.7/whatsnew/3.7.html#:~:text=the%20insertion%2Dorder%20preservation%20nature%20of%20dict%20objects%20has%20been%20declared%20to%20be%20an%20official%20part%20of%20the%20Python%20language%20spec.)

elipe17 committed:
- Remove 'test' from fixture names

raftmsohani commented on tdrs-backend/tdpservice/parsers/test/test_parse.py:
```diff
@@ -579,9 +570,9 @@ def test_parse_ssp_section1_datafile(ssp_section1_datafile, dfs):
-    expected_m1_record_count = 820
-    expected_m2_record_count = 992
-    expected_m3_record_count = 1757
+    expected_m1_record_count = 818
```
does this mean we have duplicate lines in this file?

raftmsohani replied:
never mind, I see we have two duplicates 👍

elipe17 replied:
It either implies there are duplicate records or are standard cat4 errors. Either or will force the entire case to be discarded.

raftmsohani commented on tdrs-backend/tdpservice/parsers/duplicate_manager.py:
```diff
@@ -0,0 +1,203 @@
+        @param is_new_max_precedence: boolean indicating if this error has the new max precedence
+        """
+        if has_precedence:
+            error = self.generate_error(
```
Can't we add line_number (row_number) to the error?

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -36,11 +36,11 @@ def parse_datafile(datafile, dfs):
         return errors
 
-    # TODO: write a test for this line
+    cat4_error_generator = util.make_generate_parser_error(datafile, None)
```
for duplicate errors, consider adding line number (row_number) to the error

elipe17 replied:
We can't add that up here. If we called `util. make_generate_parser_error` every time we had a cat4 error then we could add that info. But this function lives the entire lifecycle of the file being parsed. Would you think it better to move it down into the respective classes?

elipe17 committed:
Merge branch '2842-cat-4-remaining-s2-validators' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
- Fix failing tests due to merge

elipe17 committed:
Merge branch '2795-s1-duplicates' of https://github.com/raft-tech/TANF-app into 2966-unify-duplicate-detection

raftmsohani commented:
I did a test with `ssp_section1_datafile.txt`file, and added another duplicate trying to mimic a user adding a line three times by mistake and the dup errors were handled correctly.

elipe17 committed:
- Update message for failing test

elipe17 committed:
Merge branch '2795-s1-duplicates' of https://github.com/raft-tech/TANF-app into 2966-unify-duplicate-detection

elipe17 committed:
- Remove Vars from compose file
- Remove unnecessary checks from cat4 validator
- Updated error counts after removing check

elipe17 committed:
- fixed test

elipe17 committed:
- linting

elipe17 committed:
- move partial hash checking logic to schema/fields

elipe17 commented on tdrs-backend/tdpservice/parsers/row_schema.py:
```diff
@@ -101,6 +104,12 @@ def run_preparsing_validators(self, line, generate_error):
+        # This technically could be overriden if this applies to other fields, but as of right now, FAMILY_AFFILIATION
+        # is the only field that has `can_skip_partial` set to True.
+        if field.can_skip_partial:
+            self.should_skip_partial_dup = value in self.skip_values
```
@jtimpe let me know what you think of this solution. Not sure that I love it. I don't love it because now it seems that the hash generation itself should be moved up into the schema (which it could be). Im not sure it makes sense for the schema to hold that type of data though.

jtimpe replied:
perhaps the schema could store a function that generates a hash from the raw line or the partial hash from the parsed values (or maybe two functions `full_hash` and `partial_hash`). though i am seeing this as a bit of a rabbit hole. maybe we're better off coming back and refactoring in a separate ticket

elipe17 committed:
Merge branch '2842-cat-4-remaining-s2-validators' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
Merge branch '2795-s1-duplicates' of https://github.com/raft-tech/TANF-app into 2966-unify-duplicate-detection

elipe17 committed:
- Move fixtures to conftest.py

elipe17 committed:
- Fix failing tests

elipe17 commented on tdrs-backend/tdpservice/parsers/duplicate_manager.py:
```diff
@@ -0,0 +1,193 @@
+            if line_hash in self.record_hashes:
+                has_precedence, is_new_max_precedence = self.error_precedence.has_precedence(ErrorLevel.DUPLICATE)
+                existing_record_line_number = self.record_hashes[line_hash]
+                err_msg = (f"Duplicate record detected with record type "
```
@reitermb and @victoriaatraft could I get some UX input on the structure and content of the new error message?

elipe17 replied:
Exact duplicate

elipe17 commented on tdrs-backend/tdpservice/parsers/duplicate_manager.py:
```diff
@@ -0,0 +1,193 @@
+                has_precedence, is_new_max_precedence = self.error_precedence.has_precedence(
+                    ErrorLevel.PARTIAL_DUPLICATE)
+                existing_record_line_number = self.partial_hashes[partial_hash]
+                err_msg = (f"Partial duplicate record detected with record type "
```
@reitermb and @victoriaatraft could I get some UX input on the structure and content of the new error message?

elipe17 replied:
partial duplicate

elipe17 replied:
We can explicitly identify which fields identify it being a partial duplicate.

elipe17 replied:
Eric to provide mockups of these messages to miles/victoria to iterate on

elipe17 committed:
Merge branch '2842-cat-4-remaining-s2-validators' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
- fixed tests

elipe17 committed:
Revert "- move partial hash checking logic to schema/fields"

This reverts commit ab379ce79ea1575802aa5b508662935e3a0f2f9d.

elipe17 committed:
- Better way to do partial hash checking
- Still in schema

elipe17 committed:
- Moved hash generation into schema
- updated tests and factories to now support agnostic hash generation

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
- Updated error message
- fixed incorrect indentation on error counting

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
Merge branch '2795-s1-duplicates' of https://github.com/raft-tech/TANF-app into 2966-unify-duplicate-detection

elipe17 committed:
- move fixtures to conftest

elipe17 committed:
- fix lint

elipe17 committed:
Merge branch '2795-s1-duplicates' of https://github.com/raft-tech/TANF-app into 2966-unify-duplicate-detection

elipe17 committed:
- fixed tests from merge conflict

jtimpe reviewed (approved):
Good work!

jtimpe commented:
One last gut check @elipe17 - 36k lines were removed from `tdrs-backend/tdpservice/parsers/test/data/ADS.E2J.NDM1.TS53_fake.rollback.txt`, was that intentional?

elipe17 commented:
> One last gut check @elipe17 - 36k lines were removed from `tdrs-backend/tdpservice/parsers/test/data/ADS.E2J.NDM1.TS53_fake.rollback.txt`, was that intentional?

@jtimpe, that is an intentional deletion. Since I changed the logic for rollback/record deletion it became very import for the test `test_parse_big_s1_file_with_rollback` to start running. But it was unnecessary to have as many lines in the file as we did to test the logic. So I removed all lines that were unnecessary to test the cases covered in that test.

elipe17 committed:
Merge branch 'develop' into 2795-s1-duplicates

elipe17 committed:
- Update error creation logging to track all errors
- Updated queries to most correct syntax for aggregates function
- Updated tests

elipe17 committed:
- lint

elipe17 committed:
Merge branch 'develop' into 2795-s1-duplicates

elipe17 committed:
Merge branch '2795-s1-duplicates' of https://github.com/raft-tech/TANF-app into 2966-unify-duplicate-detection

elipe17 committed:
- move fixture

elipe17 committed:
Merge pull request #2970 from raft-tech/2966-unify-duplicate-detection

Unify Duplicate Detection Across Program Types

elipe17 committed:
- Updated aggregates query to take into account cat4 errors

elipe17 committed:
- lint

elipe17 committed:
Merge branch 'develop' into 2795-s1-duplicates

raftmsohani reviewed (approved):
Tested on local with two separate cases. Confirmed the errors, logs

andrew-jameson commented on tdrs-backend/tdpservice/parsers/aggregates.py:
```diff
@@ -1,9 +1,10 @@
 from .util import month_to_int, \
     transform_to_months, fiscal_to_calendar, get_prog_from_section
 from .schema_defs.utils import get_program_models, get_text_from_df
+from django.db.models import Q
```
```suggestion
from django.db.models import Q as Query
```

I've definitely done this in the past but was curious if you thought a renaming here might make the code lower down more intuitive.

elipe17 replied:
Yes I like that, will make the change!

elipe17 committed:
- import as Query

elipe17 committed:
Merge branch 'develop' into 2795-s1-duplicates

elipe17 committed:
- moved fixture to correct location

elipe17 committed:
- Updated cat4 duplicate errors to include fields that are duplicated

elipe17 committed:
- fix lint

elipe17 committed:
- update tests

reitermb reviewed (approved):
Synced on the following mockups & looks good! 

Here is an example of what it would look like for a record type that ends in a 2, 3, or 5 partial duplicate:

`Partial duplicate record detected with record type T5 at line 3. Record is a partial duplicate of the record at line number 2. Duplicated fields causing error: record type, reporting month and year, case number, family affiliation, date of birth, and social security number.`

This is what the message will look like for a record type that ends in a 1 or 4:

`Partial duplicate record detected with record type T5 at line 3. Record is a partial duplicate of the record at line number 2. Duplicated fields causing error: record type, reporting month and year, and case number.`

elipe17 committed:
Merge branch 'develop' into 2795-s1-duplicates

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
- update from merge conflict

elipe17 committed:
- Add additional exclusion criteria for necessary tests
- add missing fixture decorators

elipe17 committed:
Merge branch 'develop' into 2795-s1-duplicates

elipe17 committed:
- renamed parameters and classes to make more sense with their duties.
- Added much more detailed descriptions where information surrounding duplicate detection can be confused.

andrew-jameson reviewed (approved):
As covered in office hours, there were some minor stylistic changes requested as well as the inclusion of Eric's testing re: memory and time to be included in the PR. Otherwise, approved :)

elipe17 committed:
- implement TODOs that came out of OH

elipe17 committed:
- remove todo
- add warning labels to shared object

elipe17 committed:
- fix lint

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2795-s1-duplicates

elipe17 committed:
- adding some generic parsing flow docs

ADPennington committed:
Merge branch 'develop' into 2795-s1-duplicates

ADPennington committed:
Merge branch 'develop' into 2795-s1-duplicates

ADPennington commented:
@elipe17 below are my notes so far from qasp review. I've included test files as well for convenience. 

---
- have not yet checked implementation for tribal and SSP. ℹ️ 
- expected logic is in-place for Section 1 :rocket: 
- not seeing exact and part dup logic for Section 2 T5 records :question: 
- not seeing dup logic for Section 3 T6 records :question: 
- not seeing dup logic for Section 4 T7 records :question: 
- Should i expect any records associated with cat 4 errors to not be in the db? or just dup-related records? 
  - yes 
~~- is it possible to add the line number where the partial duplicate record was found in the error message? (see :grey_question:)~~
~~- `Partial duplicate record detected with record type T1 at line 5. Record is a partial duplicate of the record at line number 2. Duplicated fields causing error: record type, reporting month and year, and case number.` :grey_question:~~

---

test files:

_section1_
[2795_partdups_S1.txt](https://github.com/user-attachments/files/15874371/2795_partdups_S1.txt)
[2795_exactdups_S1.txt](https://github.com/user-attachments/files/15874372/2795_exactdups_S1.txt)

_section2_
[2795_partdups_s2.txt](https://github.com/user-attachments/files/15874377/2795_partdups_s2.txt)
[2795_exactdups_s2.txt](https://github.com/user-attachments/files/15874378/2795_exactdups_s2.txt)


_section3_
[2795_dups_s3.txt](https://github.com/user-attachments/files/15874363/2795_dups_s3.txt)
_section4_
[2795_dups_s4.txt](https://github.com/user-attachments/files/15874357/2795_dups_s4.txt)

elipe17 committed:
- Added useful debug logging to partial hash functions

elipe17 committed:
- Updated duplicate detection to allow "partial duplicates" of s3 and s4 files
- Added tests for partial dup s3/s4

ADPennington commented on tdrs-backend/tdpservice/parsers/aggregates.py:
```diff
@@ -39,22 +40,25 @@ def case_aggregates_by_month(df, dfs_status):
-    aggregate_data['rejected'] = ParserError.objects.filter(file=df).filter(case_number=None).distinct("row_number")\
-        .exclude(row_number=0).count()
+    error_type_query = Query(error_type=ParserErrorCategoryChoices.PRE_CHECK) | \
+        Query(error_type=ParserErrorCategoryChoices.CASE_CONSISTENCY)
```
@elipe17 does this assign an overall status of "rejected" if the file has any cat 4 errors, even if some records in the file are not associated with cat 4 type errors? or just reject the records that include cat 4 errors?

elipe17 replied:
@ADPennington, it does not. That query is strictly counting the number of cases that were rejected throughout the entirety of the file. Since records with Cat1 (PRE_CHECK) errors do not get serialized and entire cases with Cat4 (CASE_CONSISTENCY) errors don't get serialized, they contribute to the overall count of rejected cases. The code below (which can be found [here](https://github.com/raft-tech/TANF-app/blob/develop/tdrs-backend/tdpservice/parsers/models.py#L96)) is what determines the status of the DF.

```python
def get_status(self):
        """Set and return the status field based on errors and models associated with datafile."""
        errors = ParserError.objects.filter(file=self.datafile)

        # excluding row-level pre-checks and trailer pre-checks.
        precheck_errors = errors.filter(error_type=ParserErrorCategoryChoices.PRE_CHECK)\
                                .exclude(field_name="Record_Type")\
                                .exclude(error_message__icontains="trailer")\
                                .exclude(error_message__icontains="Unknown Record_Type was found.")

        case_consistency_errors = errors.filter(error_type=ParserErrorCategoryChoices.CASE_CONSISTENCY)

        row_precheck_errors = errors.filter(error_type=ParserErrorCategoryChoices.PRE_CHECK)\
                                    .filter(field_name="Record_Type")\
                                    .exclude(error_message__icontains="trailer")

        if errors is None:
            return DataFileSummary.Status.PENDING
        elif precheck_errors.count() > 0 or case_consistency_errors.count() > 0 or \
                self.total_number_of_records_created == 0:
            return DataFileSummary.Status.REJECTED
        elif errors.count() == 0:
            return DataFileSummary.Status.ACCEPTED
        elif row_precheck_errors.count() > 0:
            return DataFileSummary.Status.PARTIALLY_ACCEPTED
        else:
            return DataFileSummary.Status.ACCEPTED_WITH_ERRORS
```

elipe17 committed:
- updated file status to be consistent with expectations

elipe17 committed:
- Updated case_consistency validator to track which case has been validated to identify which case might need to be removed
- decoupled dup manager and case consistency val error counts
- Added more descriptive logging

elipe17 committed:
- Update tests

elipe17 committed:
- Add edge case cat4 test
- unblock util function that was causing errors

elipe17 committed:
- Updated manager to correctly delete serialized records
- Updated DFS to correctly count total records and total records created
- Updated cat4 test case to be more extensive

elipe17 committed:
- Updated test

elipe17 committed:
- fix lint

elipe17 committed:
Merge branch 'develop' into 2795-s1-duplicates

elipe17 committed:
- Add elastic specific exception handling

elipe17 committed:
- add extra case to file for duplicate detection

elipe17 committed:
- Added creation of LogEntry objects in exception blocks

elipe17 committed:
- fix lint

elipe17 committed:
- Add line number to duplicate error ParserError objects

elipe17 committed:
- Remove dup test class
- Updated to use correct error generator

elipe17 committed:
- fix lint

ADPennington commented on tdrs-backend/tdpservice/parsers/case_consistency_validator.py:
```diff
@@ -330,73 +341,66 @@ def __validate_s2_records_are_related(self):
+                    ))
+                elif closure_reason == '03' and not is_ssp:
+                    num_errors += self.__validate_case_closure_ftl(t4, t5s, (
+                        'At least one person who is HoH or spouse of HoH on case must have FTL months >=60.'
```
@elipe17 this is the full error message:

`At least one person who is head-of-household or spouse of head-of-household on case must have countable months toward time limit >= 60 since CLOSURE_REASON = 03: federal 5 year time limit.`

ADPennington commented on tdrs-backend/tdpservice/parsers/case_consistency_validator.py:
```diff
@@ -330,73 +341,66 @@ def __validate_s2_records_are_related(self):
+                        record,
+                        field='RPT_MONTH_YEAR',
+                        msg=(
+                            f'There should only be one {t4_model_name} record  '
```
this is now duplicative of the validator that produces messages like the following:

`Duplicate record detected with record type T4 at line 3. Record is a duplicate of the record at line number 2.`

ADPennington commented on tdrs-backend/tdpservice/parsers/case_consistency_validator.py:
```diff
@@ -183,76 +210,60 @@ def __validate_s1_records_are_related(self):
+                # loop through all t2s and t3s
+                # to find record where FAMILY_AFFILIATION == 1
+                num_errors += self.__validate_family_affiliation(num_errors, t1s, t2s, t3s, (
+                        f'Every {t1_model_name} record should have at least one corresponding '
```
duplicative error messages produced by the following file

[SSP_Sec1.txt](https://github.com/user-attachments/files/15990485/SSP_Sec1.txt)


```
Every M1 record should have at least one corresponding M2 or M3 record with the same RPT_MONTH_YEAR and CASE_NUMBER.
Every M1 record should have at least one corresponding M2 or M3 record with the same RPT_MONTH_YEAR and CASE_NUMBER.
```

ADPennington reviewed (changes requested):
@elipe17 this is in great shape. couple of suggestions on cat 4 error messages that either need clarity or are redundant and could be dropped. 

Testing Summary:
- expected logic is in-place for Section 1 :rocket: 
- expected logic is in-place for Section 2 :rocket: 
- expected logic is in-place for Section 3 :rocket: 
- expected logic is in-place for Section 4 :rocket: 
- cat 4 :warning:
- consistency across tribal, tanf, and ssp :rocket:

elipe17 committed:
- Updated based on review feedback

elipe17 committed:
- Update tests

elipe17 committed:
- fix lint

ADPennington reviewed (approved):
as per our async discussion, i think its worth revisiting the discussion about how many times the cat 4-related error messages appear in the error report, especially in scenarios where these are related to duplicates. this looks great! thanks @elipe17

elipe17 committed:
Merge branch 'develop' into 2795-s1-duplicates
